### ISSUE TITLE
Hairtrigger spec causing false failure?

### ISSUE THREAD

**Issue by benhalpern** [forem/forem#19944]
`spec/models/hair_trigger_spec.rb` runs a test which seeks the equality of two states which should be equal but are not.

This is the printed problem:

```ruby
"current migrations:"
[#<HairTrigger::Builder:0x00007ff651033b90 @adapter=nil, @compatibility=1, @options={:name=>"update_reading_list_document", :timing=>"BEFORE", :for_each=>"ROW", :generated=>true, :table=>"articles", :events=>["INSERT", "UPDATE"], :declarations=>"l_org_vector tsvector; l_user_vector tsvector"}, @chained_calls=[:on, :name, :before, :for_each, :declare], @errors=[["sqlite and mysql triggers may not be shared by multiple actions", :mysql, :sqlite]], @warnings=[], @actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');\n", @prepared_actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');", @prepared_name="update_reading_list_document">]
"current_triggers:"
[#<HairTrigger::Builder:0x00007ff65548a5c8 @adapter=nil, @compatibility=1, @options={:timing=>"BEFORE", :for_each=>"ROW", :generated=>true, :table=>"articles", :name=>"update_reading_list_document", :events=>["INSERT", "UPDATE"], :declarations=>"l_org_vector tsvector; l_user_vector tsvector"}, @chained_calls=[:on, :name, :before, :for_each, :declare], @errors=[["sqlite and mysql triggers may not be shared by multiple actions", :mysql, :sqlite]], @warnings=[], @actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');\n", @prepared_actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');", @prepared_name="update_reading_list_document">, #<HairTrigger::Builder:0x00007ff65548a5c8 @adapter=nil, @compatibility=1, @options={:timing=>"BEFORE", :for_each=>"ROW", :generated=>true, :table=>"articles", :name=>"update_reading_list_document", :events=>["INSERT", "UPDATE"], :declarations=>"l_org_vector tsvector; l_user_vector tsvector"}, @chained_calls=[:on, :name, :before, :for_each, :declare], @errors=[["sqlite and mysql triggers may not be shared by multiple actions", :mysql, :sqlite]], @warnings=[], @actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');\n", @prepared_actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');", @prepared_name="update_reading_list_document">, #<HairTrigger::Builder:0x00007ff65548a5c8 @adapter=nil, @compatibility=1, @options={:timing=>"BEFORE", :for_each=>"ROW", :generated=>true, :table=>"articles", :name=>"update_reading_list_document", :events=>["INSERT", "UPDATE"], :declarations=>"l_org_vector tsvector; l_user_vector tsvector"}, @chained_calls=[:on, :name, :before, :for_each, :declare], @errors=[["sqlite and mysql triggers may not be shared by multiple actions", :mysql, :sqlite]], @warnings=[], @actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');\n", @prepared_actions="NEW.reading_list_document :=\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, ''))), 'A') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))), 'B') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))), 'C') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_name, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_user_username, ''))), 'D') ||\n  setweight(to_tsvector('simple'::regconfig,\n    unaccent(\n      coalesce(\n        array_to_string(\n          -- cached_organization is serialized to the DB as a YAML string, we extract only the name attribute\n          regexp_match(NEW.cached_organization, 'name: (.*)$', 'n'),\n          ' '\n        ),\n        ''\n      )\n    )\n  ), 'D');", @prepared_name="update_reading_list_document">]
```

**Comment by mirie** [comment]
@jaw6 Can you summarize your findings and next steps?

**Comment by jaw6** [comment]
@mirie Findings are [here](https://github.com/forem/forem/pull/19957), summary:

* It looks like a problematic interaction between `hair_trigger` and another gem, `with_model`.
* [I tried to "pin" us to an updated `with_model`](https://github.com/forem/forem/pull/19957/commits/fb0d1f6ed1a064ea8245d0e3e56630edd8f7d302) that might help.
* I noticed that we often see this test failing when combined with a `validates_uniqueness` macro, which [we know is unsupported](https://github.com/thoughtbot/shoulda-matchers/issues/814).
* I also [tried to remove all the unsupported `validates_uniqueness` tests](https://github.com/forem/forem/pull/19957/commits/84f3862194df40a52c4a35fe927bc8c0c9e27df5)
* However, with the above in place, `hair_trigger_spec` still failed.

I'm unable to think of an additional reason _why_. (Or, for that matter, _why_ the `uniqueness` validation, which at worst is a false positive, would cause `hair_trigger_spec` to fail in this way.) Most likely, but still _guessing_, we probably have some other gem or "monkey patch" that `hair_trigger` and/or `shoulda-matchers` interacts badly with.

But beyond that guess, I'm totally stumped on this one. It's a frustrating situation.

Next steps: my recommendation — with some reluctance — is that we remove the `hair_trigger_spec` — at this point, it's causing us more harm than good. (Although, as it's currently x-ed out, removing it is low priority.)
